p2p/src/authenticated/discovery/actors/listener.rs

```diff
@@ -145,17 +238,187 @@ impl<E: Spawner + Clock + ReasonablyRealtime + Network + Rng + CryptoRng + Metri
             };
             debug!(ip = ?address.ip(), port = ?address.port(), "accepted incoming connection");
 
+            let prioritized_ip = {
+                let guard = self.prioritized_ips.read().await;
+                guard.get(&address.ip()).is_some_and(|set| !set.is_empty())
+            };
+
+            let subnet = subnet_of(address.ip());
+
+            // Drop the connection if the IP exceeds its rate limit
+            if self.ip_rate_limiter.check_key(&address.ip()).is_err() {
+                self.handshakes_ip_rate_limited.inc();
+                debug!(ip = ?address.ip(), "ip exceeded handshake rate limit");
+                self.prioritized_ips.write().await.remove(&address.ip());
+                continue;
+            }
+
+            if self.subnet_rate_limiter.check_key(&subnet).is_err() {
+                self.handshakes_subnet_rate_limited.inc();
+                debug!(ip = ?address.ip(), subnet = ?subnet, "subnet exceeded handshake rate limit");
+                self.prioritized_ips.write().await.remove(&address.ip());
+                continue;
+            }
+
+            // Attempt to reserve a slot for the handshake
+            let in_flight = loop {
+                let result = self.in_flight_handshakes.fetch_update(
```

Comment:
I don't love this locking but I think this is the best way to do this?

Reply:
I suppose the locking is OK in that it remains within this file (rather than spanning multiple actors) 🤔